### DISCUSSION TITLE
Unreliable QUIC extenstion: client and server examples

### DISCUSSION THREAD

**Discussion by nEghbal** [ngtcp2/ngtcp2#1233] [General]
I am interested in using the QUIC datagram frames instead of stream frames but all of the examples are HTTP client/servers and use only stream frames.

Does anyone develop a general-purpose client/server for sending data using datagrams?

My main goal is sending data without retransmissions, flow control, and ordering just like datagrams.

Thanks,
Nooshin

**Comment by sshock** [comment]
Having an example could be nice, but it's not too hard to figure out; if you read the programmers' guide you'll see that `ngtcp2_conn_writev_datagram` works quite similar to `ngtcp2_conn_writev_stream`, and for receiving you just hook into the `recv_datagram` callback.

https://nghttp2.org/ngtcp2/ngtcp2_conn_writev_datagram.html
https://nghttp2.org/ngtcp2/types.html#c.ngtcp2_recv_datagram

**Comment by nEghbal** [comment]
Thanks for your comments.

Actually I tired to implement write_datagrams() just like write_streams() in server.cc example to be able to send a file unreliably to the client. Then, as you said receiving the data with recv_datagram callback.

However, there are many details in write_streams() that I am not sure if I can keep them for write_datagrams() as well or not.
For example, we have the following conditions checked in server.cc/write_streams():

if (httpconn_ && ngtcp2_conn_get_max_data_left(conn_))

or 

on_send_blocked(...)

That are essential for reliable transfer with HTTP but not for datagrams.

My main goal is to release all the conditions for in-order reliable data transfer in the new example without any flow control, buffering, retransmission.

**Comment by nEghbal** [comment]
Perhaps working on 

test_ngtcp2_conn_writev_datagram(void)    and    test_ngtcp2_conn_recv_datagram(void) 

would be a good start to implement a simple unreliable client/server.

**Comment by nEghbal** [comment]
Or using simple_client.c as the client and try implementing a simple_server.c based on that would be better. 
Just to send messages to each other.

Does anyone has an idea that which part of simple_client.c must be changed to convert it to a server program?
Any hints would be appreciated.
